.github/actions/test-setup/action.yml

```diff
@@ -0,0 +1,19 @@
+name: Testing setup
+descriptions: Checkout the appropriate commit and build it before testing
+
+runs:
+  using: "composite"
+  steps:
+    - if: github.event_name == 'pull_request_target'  # checkout the would-be merge instead of PR branch head
+      uses: actions/checkout@v3
+      with:
+        ref: refs/pull/${{ github.event.number }}/merge
```

Comment:
See https://github.com/actions/checkout#checkout-pull-request-head-commit-instead-of-merge-commit

Reply:
Also found https://github.com/actions/checkout/issues/696 where they are talking about how the context object is cached so your approach might be better.

Reply:
We don't want to check out the pull request branch though

Reply:
Assuming the intention here is that we want to check out a simulation of the merge between the target branch and the pull request branch. If that is the case, my first comment would work but it caches the SHA and on re-runs it does not re-calculate that SHA. If the target branch has been merged into, those merged would be missed in the re-run. SO, I think your approach might solve that and would be better.